samples/csharp_dotnetcore/11.qnamaker/QnAMaker/QnABot.cs

```diff
@@ -47,27 +50,63 @@ public QnABot(BotServices services)
-        /// <returns>A task that represents the work queued to execute.</returns>
-        public async Task OnTurnAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        /// <returns>A <see cref="Task"/> that represents the work queued to execute.</returns>
+        /// <seealso cref="BotStateSet"/>
+        /// <seealso cref="ConversationState"/>
+        /// <seealso cref="IMiddleware"/>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
         {
-            if (context.Activity.Type == ActivityTypes.Message)
+            // Handle Message activity type, which is the main activity type for shown within a conversational interface
+            // Message activities may contain text, speech, interactive cards, and binary or unknown attachments.
+            // see https://aka.ms/about-bot-activity-message to learn more about the message and other activity types
+            if (turnContext.Activity.Type == ActivityTypes.Message)
             {
                 // Check QnA Maker model
-                var response = await _services.QnAServices[QnAMakerKey].GetAnswersAsync(context);
+                var response = await _services.QnAServices[QnAMakerKey].GetAnswersAsync(turnContext);
                 if (response != null && response.Length > 0)
                 {
-                    await context.SendActivityAsync(response[0].Answer);
+                    await turnContext.SendActivityAsync(response[0].Answer, cancellationToken: cancellationToken);
                 }
                 else
                 {
                     var msg = @"No QnA Maker answers were found. This example uses a QnA Maker Knowledge Base that focuses on smart light bulbs. 
                         To see QnA Maker in action, ask the bot questions like 'Why won't it turn on?' or 'I need help'.";
 
-                    await context.SendActivityAsync(msg);
+                    await turnContext.SendActivityAsync(msg, cancellationToken: cancellationToken);
+                }
+            }
+            else if (turnContext.Activity.Type == ActivityTypes.ConversationUpdate)
+            {
+                // Send a welcome & help message to the user.
+                await SendWelcomeMessageAsync(turnContext, cancellationToken);
+            }
+            else
+            {
+                await turnContext.SendActivityAsync($"{turnContext.Activity.Type} event detected", cancellationToken: cancellationToken);
+            }
+        }
+
+        /// <summary>
+        /// Greet new users as they are added to the conversation.
+        /// </summary>
+        /// <param name="turnContext">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
```

Comment:
>Optional [](start = 45, length = 8)

Way it's written, actually not optional.

Reply:
Nice catch